### PR TITLE
fix(rust): Raise if *_horizontal without inputs

### LINKED ISSUE #12093: `all_horizontal()` with no inputs raises inside `.select()` but is allowed inside `.filter()`
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl

df = pl.DataFrame({"foo": [1, 2, 3]})

df.filter(pl.all_horizontal())
# shape: (3, 1)
# ┌─────┐
# │ foo │
# │ --- │
# │ i64 │
# ╞═════╡
# │ 1   │
# │ 2   │
# │ 3   │
# └─────┘

df.select(pl.all_horizontal())
# ComputeError: expression: 'all_horizontal' didn't get any inputs
```

### Log output

_No response_

### Issue description

Side note: `min` and `max` allow empty inputs:

```python
>>> df.select(pl.min_horizontal())
shape: (0, 0)
┌┐
╞╡
└┘
```
I'm not sure if they are supposed to raise as `sum`, `any`, `all` do.

### Expected behavior

It should also raise inside `.filter()`?

### Installed versions

<details>

```
--------Version info---------
Polars:              0.19.12
Index type:          UInt32
Platform:            macOS-12.6.7-arm64-arm-64bit
Python:              3.11.6 (main, Oct  2 2023, 20:46:17) [Clang 14.0.0 (clang-1400.0.29.202)]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         <not installed>
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              2023.6.0
gevent:              <not installed>
matplotlib:          <not installed>
numpy:               1.25.0
openpyxl:            <not installed>
pandas:              2.0.3
pyarrow:             12.0.1
pydantic:            <not installed>
pyiceberg:           <not installed>
pyxlsb:              <not installed>
sqlalchemy:          <not installed>
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by reswqa**
This fixes #12093.

Currently, `all/any/sum_horizontal` will raise: `expression: '{method_name}' didn't get any inputs`. But `min/max_horizontal` does not. I'm not sure whether we should raise an error here or not, in other words whether we allow `*_horizontal` without any inputs. But I think they should at least maintain consistent behavior in this case. 

This PR tend to raise in this case, but If all of them shouldn't raise an error, then I would modify `all/any/sum_horizontal` instead. 

Another thing to note is that in the Python API they all allow no parameters to be passed in(as only have `*` args):

```python
def *_horizontal(*exprs: IntoExpr | Iterable[IntoExpr]) -> Expr:
```
So if we decide to raise, perhaps corresponding modifications should also be made here.

### CODE REVIEW

**Comment by stinodego** [comment]
I would expect `any/all/sum` to work without any inputs, as they have an identity (`false`/`true`/`0`, respectively). They are simply folds. If there is nothing to fold, you get the identity.

For `min/max`, there is no identity. So it would make sense for these to raise.

**Comment by stinodego** [comment]
Right, so this must be fixed. But not by raising an error - rather, we want to always return a single column with the identity.

**Comment by ritchie46** [comment]
> Right, so this must be fixed. But not by raising an error - rather, we want to always return a single column with the identity. At least, I think that's the way to go.

Yes, but that is only if the identity exist. So for `min` and `max` we can either error or return a `null` column. Rust returns `Option<T>` for `min`/`max`, so the null column might be reasonable.

**Comment by ritchie46** [comment]
The height normally is dependent of the input. Only we don't have an input atm. What normally works is returning a single value, as that will be broadcasted to the result height of other expressions. 

Now it gets a bit more complicated if we have empty dataframes. As we cannot hallucinate rows that don't exists. 

In this code `pl.EmptyDataFrame().select(pl.sum_horizontal([]).alias("foo"))`, I expect a result column `"foo"` with 0 rows.

I am not entirely sure how we will be able to achieve this, as this: `pl.EmptyDataFrame().select(pl.sum_horizontal([]).alias("foo"), pl.lit(1).alias("bar"))` should return a `DataFrame` with 1 row.

**Comment by ritchie46** [comment]
I think we should error. After thinking about this a bit, the identity is unknown at this point.

The identity is not only a value, but also a number of rows, which we don't know.

Let's error with `"cannot return empty fold because the number of output rows is unknown"

File: crates/polars-lazy/src/frame/mod.rs [hunk 1]
```diff
@@ -1412,14 +1412,15 @@ impl LazyFrame {
     /// columns are considered.
     pub fn drop_nulls(self, subset: Option<Vec<Expr>>) -> LazyFrame {
         match subset {
-            None => self.filter(all_horizontal([col("*").is_not_null()])),
+            None => self.filter(all_horizontal([col("*").is_not_null()]).unwrap()),
```
Comment (ritchie46) [inline crates/polars-lazy/src/frame/mod.rs:1415]: I don't think we should unwrap. There can be a dataframe without any columns. We have try_delayed for this.

**Review by ritchie46 — CHANGES_REQUESTED** [review]

**Pull request merged**